vmm/src/acpi.rs

```diff
@@ -402,6 +402,77 @@ fn create_spcr_table(base_address: u64, gsi: u32) -> Sdt {
     spcr
 }
 
+#[cfg(target_arch = "aarch64")]
+fn create_dbg2_table(base_address: u64) -> Sdt {
+    let namespace = b"_SB_.COM1";
+    let debug_device_info_offset = 44usize;
+    let debug_device_info_len: u16 = 22 /* BaseAddressRegisterOffset */ +
+                       12 /* BaseAddressRegister */ +
+                       4 /* AddressSize */ +
+                       namespace.len() as u16 + 1 /* zero-terminated */;
+    let tbl_len: u32 = debug_device_info_offset as u32 + debug_device_info_len as u32;
+    let mut dbg2 = Sdt::new(*b"DBG2", tbl_len, 0, *b"CLOUDH", *b"CHDBG2  ", 1);
+
+    /* OffsetDbgDeviceInfo */
+    dbg2.write_u32(36, 44);
+    /* NumberDbgDeviceInfo */
+    dbg2.write_u32(40, 1);
+
+    /* Debug Device Information structure */
+    /* Offsets are calculated from the start of this structure. */
+    let namespace_offset = 38u16;
+    let base_address_register_offset = 22u16;
+    let address_size_offset = 34u16;
+    /* Revision */
+    dbg2.write_u8(debug_device_info_offset, 0);
+    /* Length */
+    dbg2.write_u16(debug_device_info_offset + 1, debug_device_info_len);
+    /* NumberofGenericAddressRegisters */
+    dbg2.write_u8(debug_device_info_offset + 3, 1);
+    /* NameSpaceStringLength */
+    dbg2.write_u16(debug_device_info_offset + 4, namespace.len() as u16 + 1);
+    /* NameSpaceStringOffset */
+    dbg2.write_u16(debug_device_info_offset + 6, namespace_offset);
+    /* OemDataLength */
+    dbg2.write_u16(debug_device_info_offset + 8, 0);
+    /* OemDataOffset */
+    dbg2.write_u16(debug_device_info_offset + 10, 0);
+    /* Port Type */
+    dbg2.write_u16(debug_device_info_offset + 12, 0x8000);
+    /* Port Subtype */
+    dbg2.write_u16(debug_device_info_offset + 14, 0x0003);
+    /* Reserved */
+    dbg2.write_u16(debug_device_info_offset + 16, 0);
+    /* BaseAddressRegisterOffset */
+    dbg2.write_u16(debug_device_info_offset + 18, base_address_register_offset);
+    /* AddressSizeOffset */
+    dbg2.write_u16(debug_device_info_offset + 20, address_size_offset);
+    /* BaseAddressRegister */
+    dbg2.write(
+        debug_device_info_offset + base_address_register_offset as usize,
+        GenericAddress::mmio_address::<u8>(base_address),
+    );
+    /* AddressSize */
+    dbg2.write_u32(
+        debug_device_info_offset + address_size_offset as usize,
+        0x1000,
+    );
+    /* NamespaceString, zero-terminated ASCII */
+    let mut k = 0;
+    while k < namespace.len() {
```

Comment:
Sorry for not having noticed this earlier, but this code is not idiomatic.

You can use `for b in namespace.iter()` to access each byte. That drops the need for having a mutable `k` here.

Then the last `0` can be written with `debug_device_info_offset + namespace_offset as usize + namespace.len()`.

Reply:
Thanks, i'll check this. It'd still need the index, but i guess it can be solved.

Reply:
Do you still need the index? `k` only shows up in the loop and the statement to write the last `0`.

Reply:
Ah, I see. If you can't find a way to get rid of `k`, don't bother.

Reply:
You can do `for (k, b) in namespace.iter().enumerate()` instead (I don't remember exactly which position is the index, which is the value).

Reply:
From what you wrote `for (k, b) in namespace.iter().enumerate()`, `k` is the index.

Reply:
Yep, solved this way. Now everything is local to the loop.

Thanks